### PR TITLE
Scilla External Library Support

### PR DESCRIPTION
**Pull request by kaikawaliu**
## Description

This PR integrates Zilliqa Core Protocol with Scilla Interpreter for supporting contract referring to external library account. So an contract can make use of definition in external libraries instead of defining in their own contract.

Notices:
- The library contract is basically a contract which only have function or variable definition but no triggerable transitions. It's also deployed via contract transaction with proper code and data field.
- Library contract has a new field "_library" in their init data with Boolean type.
- Contract referring to external libraries will have an extra field "_extlibs" in their init data
- Besides of basic feature implementation. Library code and init data was cached in filesystem once it has been referred by any contract which can save the computation power of extracting data from data base.

Reference: 
https://github.com/Zilliqa/scilla-docs/pull/77
https://github.com/Zilliqa/scilla/wiki/Library-import-mechanism-in-Scilla

## Backward Compatibility

- [ ] This is not a breaking change
- [x] This is a breaking change

### Implementation

- [x] **ready for review**

### Integration Test (Core Team)

- [x] local machine test

- [ ] small-scale cloud test

### CODE REVIEW

File: src/libData/AccountData/Account.h [hunk 1]
```diff
@@ -135,10 +142,15 @@ class Account : public AccountBase {
   Account(const uint128_t& balance, const uint64_t& nonce,
           const uint32_t& version = ACCOUNT_VERSION);
 
+  bool GetIsLibrary(bool& is_library);
+
+  bool GetScillaVersion(uint32_t& scilla_version);
```
Comment (ansnunez) [inline src/libData/AccountData/Account.h]: Unused function. Ok to remove?

File: src/libData/AccountData/Account.h [hunk 2]
```diff
@@ -135,10 +142,15 @@ class Account : public AccountBase {
...
+
+  bool GetExternalLibs(std::vector<Address>& extlibs);
```
Comment (ansnunez) [inline src/libData/AccountData/Account.h]: Unused function. Ok to remove?

File: src/libData/AccountData/Account.cpp [hunk 3]
```diff
@@ -193,9 +191,117 @@ bool Account::DeserializeBase(const bytes& src, unsigned int offset) {
...
+        }
+      }
+    } else {
+      LOG_GENERAL(WARNING, "Wrong data format spotted");
+      return false;
+    }
+  }
+
+  if (!found_scilla_version) {
+    LOG_GENERAL(WARNING, "Didn't found scilla_version in init data");
```
Comment (ansnunez) [inline src/libData/AccountData/Account.cpp]:
```suggestion
    LOG_GENERAL(WARNING, "Didn't find scilla_version in init data");
```

File: src/libData/AccountData/Account.cpp [hunk 4]
```diff
@@ -359,49 +447,88 @@ const bytes Account::GetCode() const {
...
   }
+  is_library = m_is_library;
   scilla_version = m_scilla_version;
+  extlibs = m_extlibs;
   return true;
 }
 
+bool Account::RetrieveContractAuxiliaries() {
+  if (!isContract()) {
+    LOG_GENERAL(WARNING, "Not a contract why call GetScillaVersion");
```
Comment (ansnunez) [inline src/libData/AccountData/Account.cpp]:
```suggestion
    LOG_GENERAL(WARNING, "Not a contract");
```

File: src/libData/AccountData/AccountStoreSC.tpp [hunk 5]
```diff
@@ -649,9 +698,81 @@ Json::Value AccountStoreSC<MAP>::GetBlockStateJson(
...
+                            extlibs_exports) -> bool {
+    // export extlibs
+    for (const auto& libAddr : extlibs) {
+      if (extlibs_exports.find(libAddr) != extlibs_exports.end()) {
+        continue;
+      }
+
+      Account* libAcc = this->GetAccount(libAddr);
+      if (libAcc == nullptr) {
+        LOG_GENERAL(WARNING, "libAcc: " << libAddr << " is not existing");
```
Comment (ansnunez) [inline src/libData/AccountData/AccountStoreSC.tpp]:
```suggestion
        LOG_GENERAL(WARNING, "libAcc: " << libAddr << " does not exist");
```

File: src/libData/AccountData/AccountStoreSC.tpp [hunk 6]
```diff
@@ -729,6 +867,25 @@ bool AccountStoreSC<MAP>::ExportContractFiles(Account& contract) {
     os << DataConversion::CharArrayToString(contract.GetInitData());
     os.close();
 
+    for (const auto& extlib_export : extlibs_exports) {
```
Comment (vaivaswatha) [inline src/libData/AccountData/AccountStoreSC.tpp]: This piece of code looks to be the same as the one in `ExportCreateContractFiles`. Would it be a good idea to combine them into a function? (feel free to ignore this suggestion).

**Pull request merged**